Fix incorrect results from floating-point division (fld, cld, div)

lvlte opened the pull request:
This fixes #49450 (more details in [this comment](https://github.com/JuliaLang/julia/issues/49450#issuecomment-3694946121)).

lvlte committed:
Fix incorrect results from floating-point division (fld, cld, div)

lvlte committed:
fix whitespace check error (missing trailing newline)

adienes commented:
unfortunately does not solve the whole issue

```
julia> x = 1.125899906841625e15
1.125899906841625e15

julia> y = 0.1
0.1

julia> div(x, y, RoundDown)
1.125899906841625e16

julia> div(big(x), big(y), r)
1.1258999068416249e16
```

lvlte commented:
@adienes @Seelengrab Float64 cannot represent the value 1.1258999068416249e16. 

```julia
julia> x, y = 1.125899906841625e15, 0.1
(1.125899906841625e15, 0.1)

julia> prevfloat(x/y), x/y, nextfloat(x/y)
(1.1258999068416248e16, 1.125899906841625e16, 1.1258999068416252e16)
```

As already mentioned in the issue :

> If eps(x/y) > 1, x/y rounds to an unsafe integer which can't be floored/ceiled if it needs to because x/y ± 1 is not representable (except for 2^53 - 1).

lvlte commented:
You are missing the point : the floating-point division `x/y` returns an integer, so for consistency the fld and cld for the same type should give the same result. Using more precision should remain at the discretion of the user when using these functions. I don't think we should tweak every operation result considering the same operation with types allowing more precision. This would be like trying to fix `x - x % y` (which equals `x` in Float64) with `(x ± eps(x)) - x % y)` because the result is different with BigFloat. 

I'm not comfortable with the idea that `abs(cld(x, y) - fld(x, y))` equals 2 (in this example) as it would imply for all cases where `eps(x/y) > 1` that the absolute difference should be equal to `eps(x/y)`, which would make the results dependent on the magnitude of x/y, instead of having a consistent value that is either 0 or 1.

You didn't explain why 1.1258999068416248e16 should be preferred, and what is violating the rounding conditions.

adienes commented:
here's what I'm looking at:
```
julia> x, y
(1.125899906841625e15, 0.1)

julia> big(x) / big(y)
1.125899906841624937500000000055459109526978519729750209567102918209446384572968e16
```

so to my eyes the "correct" integer would be `11258999068416249` . but this is not representable in `Float64`. the reason I suggest to round downwards to `11258999068416248` rather than upwards to `11258999068416250`  is that is usually what I would assume flooring division (aka `RoundDown`) does



as a side note: 
> You are missing the point ... You didn't explain

I would invite you to use phrasing that more clearly reaffirms that we are collaborating together to solve the bug, rather than that we are arguing 🙂

adienes commented:
well I can appreciate the fervent defense and I do apologize for having given off the wrong impression during review.

this might be a fun one to discuss on triage, so I'll tag for review on the next meeting

lvlte commented:
The comment you mention doesn't involve x,y such that `eps(x/y) > 1`, but yes it reminds us that having `x/y` rounded to an integer doesn't mean the mathematical result is an integer. To better illustrate my point, i would like to insist on the consistency between the quotient and remainder, and between the operations that involve them : 

Example 1 `eps(x/y) ≤ 1` (no disagreement, this PR just fixes inexact results) :
```julia
julia> x, y = 1.0, 0.1
(1.0, 0.1)

julia> a, r = divrem(x, y, RoundDown)    # accurate
(9.0, 0.09999999999999995)

julia> a + r/y == x/y, a + r/y, x/y      # accurate, consistent
(true, 10.0, 10.0)
```

Example 2 `eps(x/y) > 1` (current behavior, which this PR currently doesn't affect) : 
```julia
julia> x, y = 1.125899906841625e15, 0.1
(1.125899906841625e15, 0.1)

julia> a, r = divrem(x, y, RoundDown)    # inaccurate (intrinsic roundoff)
(1.125899906841625e16, 0.03750000000005546)

julia> a + r/y == x/y, a + r/y, x/y      # inaccurate (intrinsic roundoff), consistent
(true, 1.125899906841625e16, 1.125899906841625e16)
```

Now from example 2, let say because the remainder is not zero we force `a` to be rounded further down to the previous representable integer :
```julia
julia> a = 1.1258999068416248e16         # inaccurate (arbitrary roundoff)
1.1258999068416248e16

julia> a + r/y == x/y, a + r/y, x/y      # inaccurate (arbitrary roundoff), inconsistent
(false, 1.1258999068416248e16, 1.125899906841625e16)
```

lvlte commented:
Regarding the behavior of `div` with Float16/Float32, I don't think the partial fix actually makes the decision of forcing rounding further down/up in those situations where `eps(x/y) > 1` and where the remainder is not significant. I mean, currently `div()` might behave like that with the partial fix in certain cases, but not in others (which is problematic, I'm just discovering this), eg. if we compute fld as `Float32(round(Float64(x) / Float64(y), RoundDown))` step by step, we get :
```julia
julia> x, y = 2.096153f6, 0.1f0
(2.096153f6, 0.1f0)

julia> round(Float64(x) / Float64(y), RoundDown)
2.0961529e7

julia> Float32(2.0961529e7)    # rounded down
2.0961528f7
```
```julia
julia> x, y = nextfloat(x), y + 2*eps(y)
(2.0961531f6, 0.10000002f0)

julia> round(Float64(x) / Float64(y), RoundDown)
2.0961527e7

julia> Float32(2.0961527e7)    # rounded up
2.0961528f7
```

The rounding for non-representable numbers is done by `Float32()` here and it can go both ways if not set explicitly. If the decision was made to round down/up further for non representable numbers, we should have a partial fix that makes the "final" rounding explicit : 
```julia
div(x::Float32, y::Float32, r::RoundingMode) = Float32(round(Float64(x) / Float64(y), r), r)
div(x::Float16, y::Float16, r::RoundingMode) = Float16(round(Float32(x) / Float32(y), r), r)
```

lvlte committed:
Generic fix for floating-point division (fld, cld, div)

DilumAluthge committed:
Make `CLAUDE.md` just include `@AGENTS.md` (#60722)

As suggested in
https://github.com/JuliaLang/julia/pull/60719#issuecomment-3760653690.

Replacement for #60719 (closes #60719).

(cherry picked from commit 5cbacfcf58ed41ff91a961a4f0dcc0e999ce4063)

lvlte committed:
Merge branch 'master' into floating-point_division_fix

oscardssmith commented:
Triage approves of the semantics as implimented, but would prefer a warning in the docs about the subtitles of the behavior.

nsajko commented on test/numbers.jl:
```diff
@@ -4,6 +4,8 @@ using Base.MathConstants
 using LinearAlgebra
 
+using Test
+
```
Avoid unrelated change?

```suggestion
```

lvlte committed:
Merge branch 'master' into floating-point_division_fix

lvlte committed:
docs: update div/cld/fld docstrings

- Add a note about the exactness of results with floating-point
  numbers.
- Put already existing note about floating-point division in the
  same block.

lvlte committed:
Remove unrelated change

lvlte committed:
Merge branch 'master' into floating-point_division_fix

lvlte committed:
docs: use less pedantic/opinionated wording

adienes commented on CLAUDE.md:
this file seems unrelated

lvlte committed:
Merge branch 'master' into floating-point_division_fix

lvlte committed:
More tests for floating-point div/fld/cld

lvlte committed:
docs: use more descriptive / less confusing terms

adienes commented:
thank you! in my opinion the new docs are more clear.

just need to remove the diff to `CLAUDE.md` and it looks good to go to me!

lvlte committed:
Revert changes to CLAUDE.md

lvlte committed:
docs: Fix typo in cld jldoctest
